controllers/atlasmigration_controller.go

```diff
@@ -0,0 +1,285 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	dbv1alpha1 "github.com/ariga/atlas-operator/api/v1alpha1"
+	"github.com/ariga/atlas-operator/internal/atlas"
+)
+
+// AtlasMigrationReconciler reconciles a AtlasMigration object
+type AtlasMigrationReconciler struct {
+	client.Client
+	CLI    CLI
+	Scheme *runtime.Scheme
+}
+
+// AtlasMigrationHCLTmplData is the data used to render the HCL template
+// that will be used for Atlas CLI
+type AtlasMigrationHCLTmplData struct {
+	URL       string
+	Migration struct {
+		Dir string
+	}
+	Cloud struct {
+		URL     string
+		Token   string
+		Project string
+	}
+	Data struct {
+		RemoteDir struct {
+			Name string
+			Tag  string
+		}
+	}
+}
+
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=db.atlasgo.io,resources=atlasmigrations/finalizers,verbs=update
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the AtlasMigration object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *AtlasMigrationReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	var (
+		am  dbv1alpha1.AtlasMigration
+		err error
+	)
+
+	// Get AtlasMigration resource
+	if err = r.Get(ctx, req.NamespacedName, &am); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// At the end of reconcile, update the status of the resource base on the error
+	defer func() {
+		if err != nil {
+			log.Error(err, "failed to reconcile")
+		}
+
+		clientErr := r.updateResourceStatus(ctx, am, err)
+		if err != nil {
+			log.Error(clientErr, "failed to update resource status")
+		}
+	}()
+
+	// Skip if atlas version already up to date
+	if am.Status.LastAppliedVersion == am.Spec.Version {
+		return ctrl.Result{}, nil
+	}
+
+	// Reconcile given resource
+	am.Status, err = r.reconcile(ctx, am)
+	if err != nil {
```

Comment:
if clientErr?

Reply:
This logic was moved to the next PR, we will discuss on that